app/assets/javascripts/admin/addon/services/admin-search-data-source.js

```diff
@@ -0,0 +1,317 @@
+import { tracked } from "@glimmer/tracking";
+import Service, { service } from "@ember/service";
+import { adminRouteValid } from "discourse/lib/admin-utilities";
+import { ajax } from "discourse/lib/ajax";
+import escapeRegExp from "discourse/lib/escape-regexp";
+import getURL from "discourse/lib/get-url";
+import PreloadStore from "discourse/lib/preload-store";
+import { ADMIN_NAV_MAP } from "discourse/lib/sidebar/admin-nav-map";
+import I18n, { i18n } from "discourse-i18n";
+
+// TODO (martin) Move this to javascript.rake constants, use on server too
+export const RESULT_TYPES = ["page", "setting", "theme", "component", "report"];
+const SEPARATOR = ">";
+const MIN_FILTER_LENGTH = 2;
+const MAX_TYPE_RESULT_COUNT_LOW = 15;
+const MAX_TYPE_RESULT_COUNT_HIGH = 50;
+
+export default class AminSearchDataSource extends Service {
+  @service router;
+  @service siteSettings;
+
+  plugins = {};
+  pageMapItems = [];
+  settingMapItems = [];
+  themeMapItems = [];
+  componentMapItems = [];
+  reportMapItems = [];
+  settingPageMap = {
+    categories: {},
+    areas: {},
+  };
+  @tracked _mapCached = false;
+
+  get isLoaded() {
+    return this._mapCached;
+  }
+
+  async buildMap() {
+    if (this._mapCached) {
+      return;
+    }
```

Comment:
I suspect we can use Ember [`@cached`](https://api.emberjs.com/ember/release/functions/@glimmer%2Ftracking/cached) here. It will cache a getter until one of the properties used inside mutates.